library/src/main/scala/scala/scalajs/macroImpls/UseAsMacros.scala

```diff
@@ -0,0 +1,304 @@
+           * due to named parameters.
+           */
+          val pinit = params.init
+          val noLastDefault = pinit :+ params.last.copy(isDefault = false)
+
+          find(jsMem.copy(params = pinit)) &&
+          find(jsMem.copy(params = noLastDefault))
+
+        case _ => false
+      }
+
+      find(jsMem) || hasConformingOverloads
+    }
+
+    /** Members that a facade type defines */
+    private def rawJSMembers(tpe: Type): Iterable[(String, JSMember)] = {
+
+      def isValidMember(mem: Symbol) = (
+          !JSObjectAncestors(mem.owner) &&
+          !mem.isConstructor &&
+          mem.isMethod &&
+          !mem.asTerm.isParamWithDefault
+      )
+
+      for (mem <- tpe.members if isValidMember(mem)) yield {
+        val (name, explicit) = jsName(mem.asMethod)
+
+        // apply is a special case, since when it is not explicitly specified,
+        // it will result in an actual function application in JavaScript
+        // Scala defined types cannot support this --> fail
+        if (name == "apply" && !explicit) {
+          c.abort(c.enclosingPosition, s"$tpe defines an apply method. This " +
+            "cannot be implemented by any Scala exported type, since it " +
+            "would need to chain Function's prototype.")
+        }
+
+        (name, jsMemberFor(tpe, mem.asMethod))
+      }
+    }
+
+    /** Returns the JavaScript name of a method and whether it is specified
+     *  explicitly or not.
+     */
+    private def jsName(sym: MethodSymbol): (String, Boolean) = {
+      val annot = memberAnnotations(sym).find(annotIs(_, JSNameAnnotation))
+      val name = annot.flatMap(annotStringArg).getOrElse(defaultName(sym))
+      (name, annot.isDefined)
+    }
+
+    /** Returns all exported members of a type */
+    private def exportedMembers(tpe: Type): Map[String, List[JSMember]] = {
+      val exports = tpe.baseClasses.flatMap(exportedDecls(tpe, _))
+
+      // Group exports by name
+      val lists = for {
+        (name, elems) <- exports.groupBy(_._1)
+      } yield {
+        (name, elems.map(_._2).toList)
+      }
+
+      lists.toMap
+    }
+
+    /** All exported declarations of a class.
+     *  (both @JSExportAll and @JSExport)
+     */
+    private def exportedDecls(origTpe: Type, sym: Symbol) = {
+      require(sym.isClass)
+
+      val exportAll = sym.annotations.exists(annotIs(_, JSExportAllAnnotation))
+
+      for {
+        mem <- sym.info.decls if mem.isMethod && !mem.isConstructor
+        name <- exportNames(mem.asMethod, exportAll)
+      } yield {
+        (name, jsMemberFor(origTpe, mem.asMethod))
+      }
+    }
+
+    /** Get the JS member for a method in [[origTpe]] */
+    private def jsMemberFor(origTpe: Type, sym: MethodSymbol): JSMember = {
+      sym.info.asSeenFrom(origTpe, sym.owner) match {
+        case MethodType(List(param), returnType)
+          if returnType.typeSymbol == definitions.UnitClass &&
+             sym.name.decodedName.toString.endsWith("_=") =>
```

Comment:
I guess `sym.name.endsWith(nme.SETTER_SUFFIX)` does not exist in the macro reflect API?